Add support for specifying maxLines for Text.

mpcomplete opened the pull request:
Overflow handling works with clipping, adding an ellipsis on the last
line, or fading the last line.

Fixes https://github.com/flutter/flutter/issues/7271

mpcomplete commented:
See also https://github.com/flutter/engine/pull/3338. I still need tests, but it's in a reviewable state.

abarth commented on packages/flutter/lib/src/painting/text_painter.dart:
```diff
@@ -95,6 +100,32 @@ class TextPainter {
+    _maxLines = value;
+    _paragraph = null;
+    _needsLayout = true;
+  }
```
Why is this a property on the text painter?  I would have expected it to come in through the TextStyle only

abarth replied:
Nevermind.  This is correct.

abarth commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -110,8 +113,20 @@ class RenderParagraph extends RenderBox {
+
   void _layoutText({ double minWidth: 0.0, double maxWidth: double.INFINITY }) {
-    bool wrap = _softWrap || _overflow == TextOverflow.ellipsis;
+    bool wrap = _softWrap || _overflow == TextOverflow.ellipsis; // @@@MP suspect
```
`@@@MP` ?  :)

mpcomplete replied:
I found the problem with this line.

If I have `Text(longLine, softWrap: false, maxLines: 3, overflow: TextOverflow.clip)`, then it will remain one 1 line (unless there are explicit line breaks). But if I have `Text(longLine, softWrap: false, maxLines: 3, overflow: TextOverflow.ellipsis)` then it will wrap to exactly 3 lines and truncate with an ellipsis there.

I've changed it to only force wrap if maxLines is null. This is weird and subtle, but I think is the best compromise. It allows you to say both "I don't care how many lines this takes up, but end it with an ellipsis if it exceeds the width" [1] and also "I want this to not exceed N lines, and end with an ellipsis if it exceeds N lines." [2]

[1] softWrap = false, maxLines = null
[2] softWrap = true, maxLines = N

abarth commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -180,33 +195,46 @@ class RenderParagraph extends RenderBox {
-    size = constraints.constrain(textSize);
+    final double constrainedHeight =
+      _maxLines == null ? textSize.height :
+        textSize.height.clamp(0.0, _textPainter.preferredLineHeight * _maxLines);
```
Shouldn't the text painter have taken this into account already?  If you give the text painter a max lines, can it give you back a size that's bigger than that?

mpcomplete replied:
As implemented, it can. It only uses maxLines to decide where to put the ellipsis (if any). The actual clipping is done here.

abarth replied:
We probably want to teach it to stop after that number of lines.  Consider the case where the next line has a really tall glyph that overlaps the previous line.  If you clip here, the part of the glyph that overlaps the previous line will be drawn, which isn't what we want.

abarth replied:
Specifically, I'd expect the engine to stop making line boxes after it hit the max line.

mpcomplete replied:
Just tried this, and it breaks fade overflow. If the textPainter stops after maxLines, we don't know whether we've overflowed, so we can't apply the fade.

abarth commented on packages/flutter/lib/src/painting/text_painter.dart:
```diff
@@ -180,8 +211,9 @@ class TextPainter {
+        maxLines: _maxLines,
       );
-      paragraphStyle ??= new ui.ParagraphStyle();
+      paragraphStyle ??= new ui.ParagraphStyle(maxLines: _maxLines);
```
I don't understand why `maxLines: _maxLines` is here but the other properties, like `ellipsis` aren't here.

mpcomplete replied:
Me neither. Seems I should add all the properties.

mpcomplete commented:
Added some tests which illuminate some of the weirder layout cases. (Of note are lines 114, 121, and 133.)

mpcomplete commented:
Pending travis and LGTM, this is ready for commit.

abarth commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -180,33 +194,48 @@ class RenderParagraph extends RenderBox {
     final Size textSize = _textPainter.size;
-    size = constraints.constrain(textSize);
+    final double effectiveLines =
+      // (overflow == TextOverflow.ellipsis && maxLines == null) ? 1.0 :
```
Please remove commented out code.

abarth commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -180,33 +194,48 @@ class RenderParagraph extends RenderBox {
+      // (overflow == TextOverflow.ellipsis && maxLines == null) ? 1.0 :
+      (maxLines?.toDouble() ?? double.INFINITY);
+    final double constrainedHeight =
+      textSize.height.clamp(0.0, _textPainter.preferredLineHeight * effectiveLines);
```
I don't think this is correct.  Not every line will have a height that matches the preferred line height.  It would probably be better to not to add support for a vertical fade overflow effect.

mpcomplete replied:
TextOverflow.clip relies on this code as well. I could change the engine again to give us an accurate height value for maxLines. (I'm thinking basically the same code path as TextOverflow.ellipsis without appending the ellipsis.)

IMO fade looks good as long as the height value is correct, so I think we should keep it if we can.

abarth replied:
Does the engine not give us an accurate value for the height?  I wouldn't expect clip to require and code on the framework side.

mpcomplete replied:
The engine gives us a value for the full height, ignoring maxLines. Unless we're using TextOverflow.ellipsis - then the height is clipped to the line that was ellipsized.

abarth replied:
It's probably cleaner if the engine only touches the first `maxLines` regardless of the text overflow.  Then we can keep the clip and fade effects without needing to mess with preferred line height.

mpcomplete replied:
Yeah, I agree. I'll try that. (I think I tried this before and decided against it for some reason, but I forget why. Trying again...)

mpcomplete replied:
Oh right. If the engine returns the height of the first maxLines, we can't know whether the text overflowed. But that only matters for `fade`, so, I guess I'll just remove support for that if maxLines is set.

mpcomplete replied:
OK, went with @Hixie 's suggestion and added new getter: https://github.com/flutter/engine/pull/3342

abarth commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -180,33 +194,48 @@ class RenderParagraph extends RenderBox {
+            _overflowShader = new ui.Gradient.linear(
+              <Point>[new Point(0.0, fadeStart), new Point(0.0, fadeEnd)],
+              <Color>[const Color(0xFFFFFFFF), const Color(0x00FFFFFF)]
+            );
```
Let's skip the vertical fade.  It's even less clear how it should work than the horizontal fade.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

Hixie commented:
We can also just pass a bit back from the engine which says "there was more vertical content but I ignored it because I hit the max lines"

mpcomplete committed:
Add support for specifying maxLines for Text.

Overflow handling works with clipping, adding an ellipsis on the last
line, or fading the last line.

Fixes https://github.com/flutter/flutter/issues/7271
